consensus/core/src/core.rs

```diff
@@ -111,47 +155,196 @@ impl Core {
+        }
+
+        // explicitly add our last produced block to end result
+        let mut to_propose = vec![self.last_own_block.reference()];
+        for block_ref in ancestors.into_iter() {
+            if !references_in_block.contains(&block_ref) {
+                to_propose.push(block_ref);
+            }
+        }
+
+        assert!(!to_propose.is_empty());
+
+        to_propose
+    }
+
+    fn get_blocks(&self, _block_refs: &[BlockRef]) -> Vec<Option<VerifiedBlock>> {
+        vec![]
+    }
+
+    /// Checks whether all the leaders of the previous quorum exist.
+    /// TODO: we can leverage some additional signal here in order to more cleverly manipulate later the leader timeout
+    /// Ex if we already have one leader - the first in order - we might don't want to wait as much.
+    fn last_quorum_leaders_exist(&self) -> bool {
+        // TODO: check that we are ready to produce a new block. This will mainly check that the leaders of the previous
+        // quorum exist.
+        let quorum_round = self.threshold_clock.get_round().saturating_sub(1);
+
+        let leaders = self.leaders(quorum_round);
+        let leaders_blocks = self.blocks_at_round(quorum_round, &leaders);
+
+        // Search for all the leaders
+        leaders.iter().all(|leader| {
+            leaders_blocks
+                .iter()
+                .any(|block_ref| block_ref.author == *leader)
+        })
+    }
+
+    /// Returns the blocks at the specified round for the given authorities. The method might return more than one block
+    /// per authority in case of equivocation.
+    fn blocks_at_round(&self, round: Round, authorities: &[AuthorityIndex]) -> Vec<BlockRef> {
+        // TODO: this is super dummy for now to make it work - it will be replaced by the corresponding DagState method
+        let mut ancestors = self.pending_ancestors.clone();
+        ancestors.push_back((
+            self.last_own_block.reference(),
+            self.last_own_block.timestamp_ms(),
+        ));
+
+        ancestors
+            .into_iter()
+            .filter_map(|(b, _)| {
+                if b.round == round && authorities.contains(&b.author) {
+                    Some(b)
+                } else {
+                    None
+                }
+            })
+            .collect()
+    }
+
+    /// Returns the leaders of the provided round.
+    fn leaders(&self, round: Round) -> Vec<AuthorityIndex> {
+        // TODO: this info will be retrieved from the base committers. For now just do a simple round robin so we can
+        // use it in tests.
+        vec![AuthorityIndex::new_for_test(
+            round % self.context.committee.size() as u32,
+        )]
     }
 
     fn last_proposed_round(&self) -> Round {
         self.last_own_block.round()
     }
+
+    fn last_proposed_block(&self) -> &VerifiedBlock {
+        &self.last_own_block
+    }
+
+    fn payload_to_propose(&mut self) -> Vec<Transaction> {
```

Comment:
I agree, at least for now doesn't make much sense. I was mostly thinking of going with this approach in order to not obligate the TransactionsClient implementation cache all the pending transactions, but again probably for now there isn't immediate need for that. We'll just need to ensure the underlying channel is sizeable enough so submission doesn't stuck.

Reply:
I'll open a separate PR to refactor this since it will need some additional changes in transactions consumer/client and don't want to make this PR too big now.